### PR TITLE
Profile and Settings features

### PR DESCRIPTION
**Pull request by sebbel**
Please test code thoroughly asap @Crowducate/development

### CODE REVIEW

File: both/accounts/config.js [hunk 1]
```diff
@@ -1,3 +1,31 @@
+var pwd = AccountsTemplates.removeField('password');
```
Comment (brylie) [inline both/accounts/config.js]: Please rename this variable to be more meaningful. You can use 'camelCase' notation to allow for multiple words in the variable name.

Comment (brylie) [inline both/accounts/config.js]: Also, add a line comment above this line to describe, in plain language, the purpose of the variable.

File: both/controllers/profile.js [hunk 2]
```diff
@@ -4,5 +4,8 @@ ProfileController = AppController.extend({
   },
   waitOn: function() {
       return this.subscribe('userData');
+  },
+    onBeforeAction: function (pause) {
+    AccountsTemplates.ensureSignedIn.call(this, pause);
```
Comment (brylie) [inline both/controllers/profile.js]: Check indentation here.

File: both/controllers/profileSettings.js [hunk 3]
```diff
@@ -4,5 +4,8 @@ ProfileSettingsController = AppController.extend({
   },
   waitOn: function() {
       return this.subscribe('userData');
+  },
+    onBeforeAction: function (pause) {
```
Comment (brylie) [inline both/controllers/profileSettings.js]: Check indentation here.

File: both/controllers/teach.js [hunk 4]
```diff
@@ -5,5 +5,8 @@ TeachController = AppController.extend({
     waitOn: function() {
         // Wait for the tags to be available
         return this.subscribe('tags');
-    }
+    },
+    onBeforeAction: function (pause) {
```
Comment (brylie) [inline both/controllers/teach.js:9]: Check indentation here.

File: client/templates/course/toolbar/edit-course.html [hunk 5]
```diff
@@ -1,10 +1,15 @@
 <template name="editCourse">
     <ul class="nav nav-pills navbar-right" role="tablist">
         <li>
+        	{{#if owner}}
```
Comment (brylie) [inline client/templates/course/toolbar/edit-course.html]: Check indentation.

Comment (brylie) [inline client/templates/course/toolbar/edit-course.html]: Consider renaming this helper to 'canEditCourse'. That way the markup reads 'if can edit course', which allows for our model of collaborative editing as opposed to ownership.

Comment (brylie) [inline client/templates/course/toolbar/edit-course.html]: Also, this may be included in a separate pull request, as it is not related to profile settings.

File: client/templates/course/toolbar/edit-course.js [hunk 6]
```diff
@@ -10,3 +10,26 @@ Template.editCourse.events({
     Session.set('editingCourseId', courseId);
   }
 });
+
+Template.editCourse.helpers({
+	'owner': function() {
```
Comment (brylie) [inline client/templates/course/toolbar/edit-course.js]: This template helper is not related to the pull request  'profile settings and features', so needs to be part of a separate pull request.

File: client/templates/teach/teach.js [hunk 7]
```diff
@@ -30,7 +30,8 @@ Template.teach.events({
             author: template.find('#authorName').value, // string
             keywords: template.find('#courseKeywords').value.split(','), // split keywords to array
             published: template.find('#coursePublished').value, // string
-            about: template.find('#aboutText').value // Get the about text
+            about: template.find('#aboutText').value, // Get the about text
+            owner: currentUsername
```
Comment (brylie) [inline client/templates/teach/teach.js]: We need to plan this feature a bit better, and include it in a separate pull request. This solution only allows for one 'owner' string. Specifically, what if courses can have multiple 'collaborators' or 'collaborators'?

File: client/templates/userprofile/profile.js [hunk 8]
```diff
@@ -6,4 +6,53 @@ Template.profile.helpers({
 		var first = Meteor.user().realname.split(' ');
 		return first[0];
 	}
+});
+
+Template.profile.events({
+	'click .edit-bio': function() {
+		if (Session.get("gearclicked"))
```
Comment (brylie) [inline client/templates/userprofile/profile.js]: There are three primary concerns here:
- 'gearclicked' does not accurately describe the purpose of the session variable, 
- We are using camelCase in variable names, 
- We should not pollute the Session namespace, but rather use template level reactive variable instead.

File: client/templates/userprofile/profile.js [hunk 9]
```diff
@@ -6,4 +6,53 @@ Template.profile.helpers({
...
+		{
+			var biography = $(".editable-bio").val();
+			var origin = $("<p id='biography'></p>");
```
Comment (brylie) [inline client/templates/userprofile/profile.js]: jQuery can select by the ID: **"#biography"** here.

File: client/templates/userprofile/profile.js [hunk 10]
```diff
@@ -6,4 +6,53 @@ Template.profile.helpers({
...
+			origin.text(biography);
+			$(".editable-bio").replaceWith(origin);
+			$(".statusbio").remove();
+			Session.set("gearclicked", false);
+
+		}
+		else
```
Comment (brylie) [inline client/templates/userprofile/profile.js:22]: This file does not conform to indentation conventions for this project. Specifically, lines should be indented with two spaces, at each level of indentation.

File: client/templates/userprofile/profile.js (hunk 8 above)
Comment (brylie) [inline client/templates/userprofile/profile.js]: Also, add comments throughout this function. Each line should have a corresponding plain-language comment.

File: client/templates/userprofile/profile.js [hunk 11]
```diff
@@ -6,4 +6,53 @@ Template.profile.helpers({
...
+		{
+		var editableText = $("<textarea style='width:600px;height:75px;' class='editable-bio' /> <p class='statusbio'></p>");
+		var pHtml = $("#biography").text();
+		editableText.val(pHtml);
+		$("#biography").replaceWith(editableText);
+		Session.set("gearclicked", "true");		
+		}
+
```
Comment (brylie) [inline client/templates/userprofile/profile.js]: Cleanup whitespace.

File: client/templates/userprofile/profile.js [hunk 12]
```diff
@@ -6,4 +6,53 @@ Template.profile.helpers({
...
+		var origin = $("<p id='biography'></p>");
+		origin.text(biography);
+		$(".editable-bio").replaceWith(origin);
+		$(".statusbio").remove();
+
+		}
+
+	},
+	'keyup .editable-bio': function() {
+		var postLength = $(".editable-bio").val().length;
```
Comment (brylie) [inline client/templates/userprofile/profile.js:50]: Don't mix spaces and tabs.

File: client/templates/userprofile/profile.less [hunk 13]
```diff
@@ -6,6 +6,10 @@
 	border-width: 5px;
 }
 
+#bio-error {
+	color:red;
```
Comment (brylie) [inline client/templates/userprofile/profile.less:10]: Use two spaces for each level of indentation.

File: client/templates/userprofile/profile.less [hunk 14]
```diff
@@ -36,7 +40,7 @@
 	height:100px;
 	float:left;
 	line-height: 100px;
-	text-align: center;
+	text-align: center;
```
Comment (brylie) [inline client/templates/userprofile/profile.less]: Remove trailing whitespace.

File: client/templates/userprofile/settings.html [hunk 15]
```diff
@@ -35,7 +37,7 @@ <h2 class="userprofile-sub-title">
                        value="{{profile.username}}">
             </div>
         <div class="form-group {{#if emailerror}}has-error{{/if}}">
-            <p class="profile_errors">{{emailerror}}</p>
+            <p class="email_errors errors">{{emailerror}}</p>
```
Comment (brylie) [inline client/templates/userprofile/settings.html]: To increase readability, use one space around helpers within double handlebars:
- {{wrong}}
- {{ right }}

File: client/templates/userprofile/settings.html [hunk 16]
```diff
@@ -86,7 +91,7 @@ <h2 class="userprofile-sub-title">
 		<h2 class="userprofile-sub-title">
 				{{ mf 'userProfile-subheading-password' 'Password'}}
 		</h2>
-        <p class="profile_errors">{{perrors}}</p>
+        <p class="password_errors errors">{{perrors}}</p>
```
Comment (brylie) [inline client/templates/userprofile/settings.html]: Use spaces around helper names in handlebar helpers. Rename 'pererrors' to 'password_errors' or something more meaningful.

File: client/templates/userprofile/settings.js [hunk 17]
```diff
@@ -54,16 +55,43 @@ Template.profileSettings.events({
       {
           Session.set("realerror", "Your Realname is empty!");
       }
+      setTimeout(function() {
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Why use set timeout here? Hard-coded delays, or timeouts, may cause bugs. Instead, rely on reactive data sources to update the UI.

File: client/templates/userprofile/settings.js [hunk 18]
```diff
@@ -54,16 +55,43 @@ Template.profileSettings.events({
...
     'change #userName': function(event,template) {
-        if(template.find("#userName").value)
+        var username = template.find("#userName").value;
+        if(username)
         {
-            Session.set("usererror", "");
+          
+            if(username === Meteor.user().username)
+            {
+                Session.set("usererror", "");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Use a template level reactive variable instead of the global Session object.

File: client/templates/userprofile/settings.js [hunk 19]
```diff
@@ -54,16 +55,43 @@ Template.profileSettings.events({
...
+            }
+            else
+            { 
+                
+                Meteor.call('checkUsername', username, function(error) 
+                {
+                    if(error){
+                        Session.set("usererror", "Username taken already");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Use template level reactive variable instead of Session object.

File: client/templates/userprofile/settings.js [hunk 20]
```diff
@@ -54,16 +55,43 @@ Template.profileSettings.events({
...
+                    }
+                    else
+                    {
+                        Session.set("usererror", "");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Use Reactive Variable instead of Session.

File: client/templates/userprofile/settings.js [hunk 21]
```diff
@@ -54,16 +55,43 @@ Template.profileSettings.events({
...
+               
+            }
         }
         else
         {
             Session.set("usererror", "Your username is empty!");
         }
+
+        setTimeout(function() {
+         $('.username_errors').text("");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Why use setTimeout? This may introduce bugs.

File: client/templates/userprofile/settings.js [hunk 22]
```diff
@@ -82,26 +110,40 @@ Template.profileSettings.events({
...
+        
         var username = template.find("#userName").value;
         var gender = template.find("#gender option:selected").value;
         var language = template.find("#language option:selected").value;
         var biography = template.find("#biography").value;
         var email = template.find("#email").value;
-
+        if (biography.length > 300)
+        {
+            $("#bio-error").text("You may not have more than 300 Characters in your Biography");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: These kinds of field validation should be done with the AutoForm/Simple Schema packages.

File: client/templates/userprofile/settings.js [hunk 23]
```diff
@@ -82,26 +110,40 @@ Template.profileSettings.events({
...
+        }
+        else
+        {
+            $("#bio-error").text("");
         if (Meteor.userId())
         {
-            if (realname && username && gender && language && email && isEmail(email))
+            var nachricht = Session.get("usererror");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Switch to reactive variable. Use camelCase.

File: client/templates/userprofile/settings.js [hunk 24]
```diff
@@ -82,26 +110,40 @@ Template.profileSettings.events({
...
+            if (username && gender && language && email && isEmail(email) && (!(nachricht) || username == Meteor.user().username))
             {
                 Session.set("basicsuccess", "Data successfully changed!");
+               // setTimeout(Session.set("basicsuccess", ""),5000);
+
                 console.log("success");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Clean up console log.

File: client/templates/userprofile/settings.js [hunk 25]
```diff
@@ -115,13 +157,17 @@ Template.profileSettings.events({
             }
 
         }
+        }
 
             //Session.set("errors", "");
-            Meteor.call("User.update", Meteor.userId(),"realname", realname);
+           // Meteor.call("User.update", Meteor.userId(),"realname", realname);
 
+        setTimeout(function() {
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Please use SimpleSchema and Autoform for validations. setTimeout may introduce bugs and unnecessary delays. Rely on reactive data sources (reactive variables, etc) for UI updates.

File: client/templates/userprofile/settings.js [hunk 26]
```diff
@@ -130,19 +176,42 @@ Template.profileSettings.events({
             if (Meteor.userId()) {
             Accounts.changePassword(oldpassword,password, function(error) {
                 if(error) {
-                    console.log("Something went wrong! " + error);
+                    if(error.reason === "Match failed")
+                    {
+                        Session.set("perrors", "Your old password is empty!");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Use Reactive Variable instead of Session.

File: client/templates/userprofile/settings.js [hunk 27]
```diff
@@ -130,19 +176,42 @@ Template.profileSettings.events({
...
+                    }
+                    else if(error.reason === "Incorrect password")
+                    {
+                        Session.set("perrors", "Your old Password is incorrect!");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Use Reactive Variable instead of session.

File: client/templates/userprofile/settings.js [hunk 28]
```diff
@@ -130,19 +176,42 @@ Template.profileSettings.events({
...
+                    }
+                    else
+                    {
+                        Session.set("perrors", error.reason);
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Use Reactive Variable instead of Session.

File: client/templates/userprofile/settings.js [hunk 29]
```diff
@@ -130,19 +176,42 @@ Template.profileSettings.events({
...
                 }
             });
             }
         }
-        else {
+        else 
+        {
             Session.set("perrors", "Your passwords do not match.")
         }
+        setTimeout(function() {
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: sessionTimeout may cause bugs and unnecessary delays. Use reactive data sources to update UI. Use SimpleSchema and Autoform for field validation.

File: client/templates/userprofile/settings.js [hunk 30]
```diff
@@ -130,19 +176,42 @@ Template.profileSettings.events({
...
+         $('.password_errors').text("");
+        }, 5000);
+    },
+
+    'keyup #biography': function() {
+        var postLength = $("#biography").val().length;
+        var charactersLeft = 300 - postLength;
+        $('.statusbio').text(charactersLeft + " characters left");
```
Comment (brylie) [inline client/templates/userprofile/settings.js]: Use a more specific ID for the character counting display. Classes are used when multiple elements of the same type may be visible on screen.

File: client/templates/userprofile/settings.less [hunk 31]
```diff
@@ -13,4 +13,9 @@
 .password_success {
   background-color: darkgreen;
   color: white;
+}
+
+#bio-error {
```
Comment (brylie) [inline client/templates/userprofile/settings.less:18]: Bootstrap has a class for error text. Use existing classes, rather than creating CSS for similar effects.

File: server/methods/users.js [hunk 32]
```diff
@@ -8,6 +8,8 @@ Meteor.methods({
             switch(target) {
                 case "username":
                     console.log("Target is username");
```
Comment (brylie) [inline server/methods/users.js]: Clean up console logs.

**Comment by brylie** [comment]
This pull request needs extensive revision before merging. The primary areas of concern are:
- **whitespace** - use [two spaces](https://github.com/Crowducate/crowducate-platform/blob/develop/.jshintrc#L20) for each level of indentation and no trailing whitespace
- **variable names** - use [camelCase](https://github.com/Crowducate/crowducate-platform/blob/develop/.jshintrc#L15) for variable names
- **reactive data sources** - use Reactive Variable instead of Session for template-level reactivity
- **cruft** - clean up console logs, unnecessary whitespace (empty lines where not serving to improve code readability), etc.
- **code reuse** - use existing Bootstrap CSS classes for error text
- **form validation** - the SimpleSchema and Autoform packages handle form validation, use them instead of creating manual validation logic
- **single purpose PR** - this pull request has functionality for checking ownership of courses, which is outside of the scope of 'profile management'.

# References
- [Idiomatic JavaScript: Whitespace](https://github.com/rwaldron/idiomatic.js#whitespace)
- [Reactive Dict, Reactive Vars, and Session Variables](http://themeteorchef.com/snippets/reactive-dict-reactive-vars-and-session-variables/) - demonstrates Reactive Variable and other reactive data sources
- [Bootstrap 3: Contextual background colors](http://getbootstrap.com/css/#helper-classes-backgrounds)
- [AutoForm](https://github.com/aldeed/meteor-autoform#autoform) and [SimpleSchema](https://github.com/aldeed/meteor-simple-schema) packages

**Comment by sebbel** [comment]
Thank you for your detailed suggestions :)

**Comment by brylie** [comment]
Cheers! I can help with some of the improvements, next week. :-)

File: both/accounts/config.js (hunk 1 above)
Comment (sebbel) [inline both/accounts/config.js]: https://github.com/meteor-useraccounts/core/blob/master/Guide.md#login-with-username-or-email

Referencing for future comment

**Comment by brylie** [comment]
@sbe what IDE are you using? Check if you can install a JSHint linter plugin, which could automatically detect some of the more trivial details such as whitespace and variable naming conventions.

**Comment by sebbel** [comment]
I am currently using ~~Atom~~ Webstorm. And I managed to find the indentation tool :dancer:

File: client/templates/userprofile/settings.js (hunk 17 above)
Comment (sebbel) [inline client/templates/userprofile/settings.js]: Using set.timeout to remove errors after 5 seconds, as suggested by @Rahbaran

Comment (sebbel) [inline client/templates/userprofile/settings.js]: I'm open for any suggesting to achieve the same.

File: client/templates/course/toolbar/edit-course.js (hunk 6 above)
Comment (sebbel) [inline client/templates/course/toolbar/edit-course.js]: removed

**Comment by sebbel** [comment]
Hey @brylie, 

I hope I caught all the whitespace and identation errors, I would be grateful if you could have a quick glance. I have removed the owner/canEdit helper and if from the commit, as it is not related to this feature set.

I might need your help on reactive vars and how to refactor the error messages with usage of simpleform and bootstrap

File: client/templates/userprofile/settings.js (hunk 17 above)
Comment (brylie) [inline client/templates/userprofile/settings.js]: Use reactive variables instead of the global Session variable. I 
outlined some resources in the pull request discussion.

**Comment by brylie** [comment]
For now, start with an empty project. Build a form using SimpleSchema 
and Autoform. You can also install twbs:bootstrap in order to have 
Bootstrap 3 in your testing project.

**Comment by sebbel** [comment]
I am coming across problems with the autoform package. I managed to get the link to the current profile, so it can get edited, but when I include the email address to be edited as well, the form no longer updates the values.

http://stackoverflow.com/questions/32563172/autoform-editing-profiles

**Comment by brylie** [comment]
Take a look at how yogiben:meteor-starter handles user profiles. It is written in CoffeeScript, but can easily be converted for our project:
- [profile.html](https://github.com/yogiben/meteor-starter/blob/master/client/views/profile/profile.html)
- [users.schema:emails](https://github.com/yogiben/meteor-starter/blob/master/collections/users.coffee#L262)

**Comment by brylie** [comment]
Also, lets keep this profile feature really simple for now, since there are many things to consider such as code consistency, new concepts, etc.

**Comment by brylie** [comment]
Also, please remember to remove code relating to course ownership, as it should be in a different pull request/feature branch.

**Comment by sebbel** [comment]
@brylie I did that her already https://github.com/Crowducate/crowducate-platform/commit/51276877c0e8d3f6ba52bc8210ca8bf9ee47e3ce

**Comment by brylie** [comment]
There is a bit of cleanup left in [teach.js:34](https://github.com/Crowducate/crowducate-platform/blob/feature/profile/client/templates/teach/teach.js#L34)

**Comment by sebbel** [comment]
I have made significant progress on autoform, take a look at the code if you like. @Crowducate/development

**Comment by Rahbaran** [comment]
Looks good so far. Some remarks:
1. Please add "changing password" feature (as in the previous version)
2. Please add "add bio" feature (as in the previous version; with max. chars)
3. change to “hello USERNAME" (currently email)
4. for this version: only show created (= “canEdit”) courses (i.e.: comment out dropdown (only show canEdit)

**Comment by sebbel** [comment]
@crowducate/development i am running into problems refactoring it the 'autoform' way. 
- password should be changed using an autoform hook but I cannot submit the form anymore when I set type=normal
- on profile page I cannot update biography without supplying the other fields as they are required.

help is kindly requested.

**Comment by sebbel** [comment]
There is currently a bug, where the Error Message for the Password change is showing even on a successful change.

**Comment by brylie** [comment]
@sbe, please close this pull request until it is ready to be reviewed/merged. Working in a separate branch is sufficient to protect your work.
